src/openms/source/CHEMISTRY/ModifiedPeptideGenerator.cpp

```diff
@@ -235,131 +227,48 @@ namespace OpenMS
         // Term specificity is ANYWHERE on the peptide, C_TERM or N_TERM
         // (currently no explicit support in OpenMS for protein C-term and
         // protein N-term)
+        // TODO This is not true anymore!
         const ResidueModification::TermSpecificity& term_spec = v->getTermSpecificity();
+        // TODO why are we not checking for existing modifications like in the N/C Term case above??
         if (term_spec == ResidueModification::ANYWHERE)
         {
-          map_compatibility[static_cast<int>(residue_index)].push_back(v);
+          mod_compatibility.emplace_back(residue_index, v);
         }
+        // TODO think about if it really is the same case as the one above.
         else if (term_spec == ResidueModification::C_TERM && residue_index == (peptide.size() - 1))
         {
-          map_compatibility[C_TERM_MODIFICATION_INDEX].push_back(v);
+          mod_compatibility.emplace_back(C_TERM_MODIFICATION_INDEX, v);
         }
         else if (term_spec == ResidueModification::N_TERM && residue_index == 0)
         {
-          map_compatibility[N_TERM_MODIFICATION_INDEX].push_back(v);
+          mod_compatibility.emplace_back(N_TERM_MODIFICATION_INDEX, v);
         }
       }
     }
 
-    // Check if no compatible site that can be modified by variable
-    // modification. If so just return peptides without variable modifications.
-    const Size compatible_mod_sites = map_compatibility.size();
-    if (compatible_mod_sites == 0)
-    {
-      if (keep_unmodified)
-      {
-        all_modified_peptides.push_back(peptide);
-      }
-      return;
-    }
+    Size max_placements = std::min(max_variable_mods_per_peptide, mod_compatibility.size());
 
-    // generate powerset of max_variable_mods_per_peptide sized subset of all compatible modification sites
-    Size max_placements = std::min(max_variable_mods_per_peptide, compatible_mod_sites);
-    for (Size n_var_mods = 1; n_var_mods <= max_placements; ++n_var_mods)
+    vector<pair<unsigned, vector<AASequence>>> mod_peps_w_depth = {{0, {peptide}}};
+    for (auto& [idx, mods] : mod_compatibility)
     {
-      // enumerate all modified peptides with n_var_mods variable modified residues
-      Size zeros = std::max((Size)0, compatible_mod_sites - n_var_mods);
-      vector<bool> subset_mask;
-
-      for (Size i = 0; i != compatible_mod_sites; ++i)
+      // copy the complete sequences from last iteration
+      for (auto prev_res : mod_peps_w_depth)
       {
-        // create mask 000011 to select last (e.g. n_var_mods = 2) two compatible sites as subset from the set of all compatible sites
-        if (i < zeros)
-        {
-          subset_mask.push_back(false);
-        }
-        else
-        {
-          subset_mask.push_back(true);
-        }
+        // extend mod_peps_w_depth by adding variants with the next mod, if max_placements is not reached
+        addMods_(idx, mods, mod_peps_w_depth, max_placements, var_mods);
```

Comment:
Does this take care that one doesn't generate the same isoform twice when going to depth+1? e.g. two sites a,b can both be modified
depth1: either a modified or b modified
depth2: a modified and b modified but also b modified and a modified

Reply:
yes because you only go over the vector of mod locations once.